python/ray/serve/_private/application_state.py

```diff
@@ -0,0 +1,215 @@
+            deployment_timestamp=self.deployment_timestamp,
+        )
+
+
+class ApplicationStateManager:
+    def __init__(self, deployment_state_manager):
+        self.deployment_state_manager = deployment_state_manager
+        self._application_states: Dict[str, ApplicationState] = {}
+
+    def delete_application(self, name: str):
+        """Delete application by name"""
+        if name not in self._application_states:
+            return
+        self._application_states[name].delete()
+
+    def deploy_application(self, name: str, deployment_args: List[Dict]):
+        """Deploy single application
+
+        Args:
+            name: application name
+            deployment_args: deployment args
+        """
+        if name in self._application_states:
+            self._application_states[name].deployment_params = deployment_args
+        else:
+            self._application_states[name] = ApplicationState(
+                name,
+                self.deployment_state_manager,
+                deployment_args,
+            )
+        self._application_states[name].deploy()
+
+    def get_deployments(self, app_name: str) -> List[str]:
+        """Return all deployment names by app name"""
+        if app_name not in self._application_states:
+            return []
+        return self._application_states[app_name].get_all_deployments()
+
+    def get_deployments_statuses(self, app_name: str) -> List[DeploymentStatusInfo]:
+        """Return all deployment statuses by app name"""
+        if app_name not in self._application_states:
+            return []
+        return self._application_states[app_name].get_deployments_statuses()
+
+    def get_app_status(self, name: str) -> ApplicationStatusInfo:
+        if name not in self._application_states:
+            return ApplicationStatusInfo(
+                ApplicationStatus.NOT_EXISTED,
+                message=f"Application {name} doesn't existed",
+                deployment_timestamp=0,
+            )
+        return self._application_states[name].get_application_status_info()
+
+    def create_application_state(
+        self, name: str, deploy_obj_ref: ObjectRef, deployment_time: float = 0
+    ):
+        """Create application state
+        This is used for holding the deploy_obj_ref which is created by run_graph method
+        """
+        self._application_states[name] = ApplicationState(
+            name,
+            self.deployment_state_manager,
+            deploy_obj_ref=deploy_obj_ref,
+            deployment_time=deployment_time,
+        )
+
+    def get_deployment_timestamp(self, name: str) -> float:
+        if name not in self._application_states:
+            return -1
+        return self._application_states[name].deployment_timestamp
+
+    def update(self):
+        """Update each application state"""
+        apps_to_be_deleted = []
+        for name, app in self._application_states.items():
+            app.update()
+            if app.to_be_deleted:
```

Comment:
Why not just call `del` here? (Why do we need `apps_to_be_deleted`?)

Reply:
Ah I think this is similar to the existing `deployment_state_manager`, I don't fully remember why we needed it there but it does seem that here we could just immediately `del`